### PR TITLE
Fully-automate development setup with Gitpod

### PR DESCRIPTION
**Pull request by ghuntley**
Howdy @iluwatar! 👋

Here is a PR that fully automates the dev setup of java-design-patterns with Gitpod, providing anyone with a pre-built, browser-based development environment in one click:

[Open in Gitpod](https://gitpod.io/#https://github.com/iluwatar/java-design-patterns/pull/1788)

### What it does

* Comes with all Java dependencies pre-installed 
* Installs maven dependencies ahead of time and precompiles all of the patterns (saving everyone 21 minutes of wait time!)  

I hope this can make life easier for contributors, and for maintainers who need to review many PRs (hint: you can already open this PR in Gitpod to easily review & test it without messing up your local checkout).

### CODE REVIEW

**Comment by iluwatar** [comment]
Thanks for the PR. I don't have previous experience with Gitpod so this requires a bit of trial and consideration. I'm also inviting @ohbus to review this and share opinions.

**Review by ohbus — COMMENTED** [review]
@ghuntley As I can see this works out of the box and surely the future is a Browser-based development environment.



Your gitpod file does not include the xvfb dependency we need for running defeating the whole purpose of setting it up.

I would also want to know that how are workspaces segregated as an empty workspace for this particular repo idling on the master branch was on over 8 load average on a 16-thread machine and using 30 GiB of 34 GiB memory.

***

One more thing is that I can see you have already added the badge in our [Wiki](https://github.com/iluwatar/java-design-patterns/wiki/01.-How-to-contribute).

So one thing is that I would not want to lock our developers for cloud-based development is to only use Gitpod, as there are alternatives from GitHub itself as well like [Codespaces](https://github.com/features/codespaces), and many more will soon be emerging as developers start using these in their daily work.

**Comment by ghuntley** [comment]
> Your gitpod file does not include the xvfb dependency we need for running defeating the whole purpose of setting it up.

The `xvfb` dependency is provided by https://github.com/iluwatar/java-design-patterns/pull/1788/files#diff-454eff0561fdbb1cdf08bef7a76ddf0f82c65fca08a150c43650f5098ecf7118R24 which is installed via https://github.com/gitpod-io/workspace-images/blob/abd6818f4a9db3b2e7c7d17d4af5fdba17b0ccb4/full-vnc/Dockerfile#L3.


During the prebuild (https://www.gitpod.io/docs/prebuilds/) of the workspace `xvfb-run ./mvnw clean verify` is executed and the resulting compilation and installation of maven dependencies are cached on Gitpod's servers which means the workspace starts fast and the steps of install mvn dependencies/compilation happen ahead of time thus saving consumers/contributors/maintainers approx 21 minutes of time. 



> I would also want to know that how are workspaces segregated as an empty workspace for this particular repo idling on the master branch was on over 8 load average on a 16-thread machine and using 30 GiB of 64 GiB memory.

Workspaces are Kubernetes pods that have run-time isolation via [runc](https://github.com/opencontainers/runc). We spin up spare pods ahead of time without an owner and when a request comes through - click this link:

https://gitpod.io/#https://github.com/iluwatar/java-design-patterns/pull/1788

1. The empty pod is assigned to you
2. The context URL is used to clone your workspace
3. If a preload exists, that is restored (think CI/CD for development environments)
4. User is presented with VS Code and is ready to code.

Pods run on Kubernetes nodes with 16-threads and 64GiB of memory and the resources shared between other pods which run on the node. We have a custom Kubernetes resources scheduler that works behind the scenes to that enables developers to consume the full 16-threads and 64GiB of memory when needed whilst ensuring abusive neighbours (ie crypto miners) don't ruin the experience for all. I'm simplifying things a heap here if you want more insights 

https://www.youtube.com/watch?v=svV-uE0Cdjk


> So one thing is that I would not want to lock our developers for cloud-based development is to only use Gitpod, as there are alternatives from GitHub itself as well like [Codespaces](https://github.com/features/codespaces), and many more will soon be emerging as developers start using these in their daily work.

Gitpod is available right now for consumers, contributors and maintainers free of charge. Codespaces is still in private beta, the release date is uncertain (the release milestone has slipped three times now) and so is the pricing structure. 

The future _absolutely_ is cloud-based development because it solves so many problems such as being able to develop from anywhere on any device, automation and removal of common developer tasks (cloning, switching branches, installation of dependencies) plus the big-ticket item - supply chain trust. Instead of pulling down untrusted code (inc OSS dependencies) and executing them on your local computer, all development happens in a sandbox. It's rad.

Codespaces is rad (I'm in the beta) and when they launch you should also add them to your wiki and the readme! Over at https://github.com/ghuntley/ghuntley/blob/trunk/.devcontainer/devcontainer.json#L4 you'll find my personal experiments of sharing a single configuration between:

* vscode remote (ssh)
* vscode remote (docker)
* vscode local
* gitpod
* codespaces

> > UPDATE: I could not build the project on the workspace after checking out this commit.

Ah, that's interesting. With Gitpod you don't need to check out the commit or use "git clone/switch branches". Just click this link.

https://gitpod.io/#https://github.com/iluwatar/java-design-patterns/pull/1788

Once merged the `pull/1788` suffix will not be needed.

File: .gitpod.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+# Enable prebuilds of your project to enable faster workspace start times.
+# https://www.gitpod.io/docs/prebuilds/#configure-the-github-app
+github:
+  prebuilds:
+    master: true
+    branches: true
+    pullRequests: true
+    pullRequestsFromForks: true
+    addCheck: true
```
[the remark this answers is not in the document]
Comment (ghuntley) [reply .gitpod.yml]: ✅ resolved by adding an empty line. You can review the pull-request and test everything out by clicking on https://gitpod.io/#https://github.com/iluwatar/java-design-patterns/pull/1788

**Comment by iluwatar** [comment]
Thanks for pointing that out @ohbus. I'm terribly sorry @ghuntley but I just don't see the feature useful enough to accept and maintain. I'm reverting the change in the wiki and closing this PR.

For the future, it's always better to open an issue first to discuss the feature before jumping to the implementation.

**Pull request closed without merging**